Shift error when trying to use `mem.writeIntSliceLittle` with `u8` truncated value

## System Information

- **Zig version**: `0.9.0-dev.951+9e2472706`
- **Operating System**: Alpine Linux (Edge) running kernel `5.10.16.3-microsoft-standard-WSL2`
- **Example Usage:** https://github.com/xpyxel/cobalt/blob/dev/src/memory.zig#L15 

## Synopsis

I feel like this is a bug, as since both the type of the value, and the type given to the function match, it should properly handle writing the value to the buffer, however instead it errors with the following:

```
/home/pyxel/.local/zig/versions/0.9.0-dev.951+9e2472706/files/lib/std/mem.zig:1450:18: error: integer value 8 cannot be coerced to type 'u3'
        bits >>= 8;
                 ^
./repro.zig:8:32: note: called from here
    std.mem.writeIntSliceLittle(u8, buffer, @truncate(u8, example_variable));
                               ^
./repro.zig:3:29: note: called from here
pub fn main() anyerror!void {
                            ^
/home/pyxel/.local/zig/versions/0.9.0-dev.951+9e2472706/files/lib/std/mem.zig:1450:14: note: referenced here
        bits >>= 8;
```

To reproduce this error you can use the following code:

```zig
const std = @import("std");

pub fn main() anyerror!void {
    const allocator = std.heap.page_allocator;
    var buffer = try allocator.alloc(u8, 65535);
    defer allocator.free(buffer);
    const example_variable: u64 = 1_000_000_000;
    std.mem.writeIntSliceLittle(u8, buffer, @truncate(u8, example_variable));
}
```

--- a/lib/std/mem.zig
+++ b/lib/std/mem.zig
@@ -1439,9 +1439,13 @@ pub fn writeInt(comptime T: type, buffer: *[@divExact(@typeInfo(T).Int.bits, 8)]
 pub fn writeIntSliceLittle(comptime T: type, buffer: []u8, value: T) void {
     assert(buffer.len >= @divExact(@typeInfo(T).Int.bits, 8));
 
-    if (@typeInfo(T).Int.bits == 0)
+    if (@typeInfo(T).Int.bits == 0) {
         return set(u8, buffer, 0);
-
+    } else if (@typeInfo(T).Int.bits == 8) {
+        set(u8, buffer, 0);
+        buffer[0] = @bitCast(u8, value);
+        return;
+    }
     // TODO I want to call writeIntLittle here but comptime eval facilities aren't good enough
     const uint = std.meta.Int(.unsigned, @typeInfo(T).Int.bits);
     var bits = @bitCast(uint, value);
@@ -1459,8 +1463,13 @@ pub fn writeIntSliceLittle(comptime T: type, buffer: []u8, value: T) void {
 pub fn writeIntSliceBig(comptime T: type, buffer: []u8, value: T) void {
     assert(buffer.len >= @divExact(@typeInfo(T).Int.bits, 8));
 
-    if (@typeInfo(T).Int.bits == 0)
+    if (@typeInfo(T).Int.bits == 0) {
         return set(u8, buffer, 0);
+    } else if (@typeInfo(T).Int.bits == 8) {
+        set(u8, buffer, 0);
+        buffer[buffer.len - 1] = @bitCast(u8, value);
+        return;
+    }
 
     // TODO I want to call writeIntBig here but comptime eval facilities aren't good enough
     const uint = std.meta.Int(.unsigned, @typeInfo(T).Int.bits);
@@ -2127,6 +2136,30 @@ fn testWriteIntImpl() !void {
         0x00,
         0xAB,
         0xCD,
+    }));
+
+    writeIntSlice(u8, bytes[0..], 0x12, Endian.Big);
+    try testing.expect(eql(u8, &bytes, &[_]u8{
+        0x00, 0x00, 0x00, 0x00,
+        0x00, 0x00, 0x00, 0x12,
+    }));
+
+    writeIntSlice(u8, bytes[0..], 0x12, Endian.Little);
+    try testing.expect(eql(u8, &bytes, &[_]u8{
+        0x12, 0x00, 0x00, 0x00,
+        0x00, 0x00, 0x00, 0x00,
+    }));
+
+    writeIntSlice(i8, bytes[0..], -1, Endian.Big);
+    try testing.expect(eql(u8, &bytes, &[_]u8{
+        0x00, 0x00, 0x00, 0x00,
+        0x00, 0x00, 0x00, 0xff,
+    }));
+
+    writeIntSlice(i8, bytes[0..], -1, Endian.Little);
+    try testing.expect(eql(u8, &bytes, &[_]u8{
+        0xff, 0x00, 0x00, 0x00,
+        0x00, 0x00, 0x00, 0x00,
     }));
 }
 

